File: construction_v1/report/report_utils.py
import frappe
from frappe.utils import getdate, add_months, add_days, get_first_day

def get_period_dates(filters):
    """
    Update filters with date_from and date_to if missing, based on period_type.
    """
    if not filters.get("date_from") or not filters.get("date_to"):
        period_type = filters.get("period_type")
        date_to = getdate()
        
        if period_type == "Weekly":
            # Start of current week (Monday)
            date_from = add_days(date_to, -date_to.weekday())
        elif period_type == "Monthly":
            date_from = get_first_day(date_to)
        elif period_type == "Quarterly":
            month = date_to.month
            quarter_start_month = ((month - 1) // 3) * 3 + 1
            date_from = getdate(f"{date_to.year}-{quarter_start_month:02d}-01")
        elif period_type == "Annual":
            date_from = getdate(f"{date_to.year}-01-01")
        else:
            # Default fallback for Custom or others: last 30 days
            date_from = add_days(date_to, -29)
            
        filters.date_from = date_from
        filters.date_to = date_to
    
    return filters

File: construction_v1/public/js/report_utils.js
/**
 * Shared utility functions for Assistant CRM Reports.
 */

frappe.provide("construction_v1.report_utils");

/**
 * Common Period Type options for reports.
 */
construction_v1.report_utils.period_type_options = "Weekly\nMonthly\nQuarterly\nAnnual\nCustom";

/**
 * Handle period_type change to update date_from and date_to.
 */
construction_v1.report_utils.handle_period_change = function () {
    let period_type = frappe.query_report.get_filter_value("period_type");
    let today = frappe.datetime.get_today();
    let date_from = "";
    let date_to = today;

    if (period_type === "Weekly") {
        // Start of current week (Monday)
        // JavaScript getDay(): 0 is Sunday, 1 is Monday...
        let d = new Date();
        let day = d.getDay();
        let diff = d.getDate() - day + (day == 0 ? -6 : 1); // adjust when day is sunday
        let weekStart = new Date(d.setDate(diff));
        date_from = frappe.datetime.str_to_user(weekStart);
    } else if (period_type === "Monthly") {
        date_from = frappe.datetime.month_start();
    } else if (period_type === "Quarterly") {
        let d = new Date();
        let quarter = Math.floor(d.getMonth() / 3);
        let startMonth = quarter * 3;
        let quarterStart = new Date(d.getFullYear(), startMonth, 1);
        date_from = frappe.datetime.str_to_user(quarterStart);
    } else if (period_type === "Annual") {
        date_from = frappe.datetime.year_start();
    }

    if (date_from && period_type !== "Custom") {
        frappe.query_report.set_filter_value("date_from", date_from);
        frappe.query_report.set_filter_value("date_to", date_to);
    }
};

